### PR TITLE
Fix left click on air being ignored and right click on block being handled twice

### PR DESCRIPTION
**Pull request by Yeregorix**
### Left click on air

**Bukkit**: Left click on air was already detected properly.
**Sponge**: Added a listener for `InteractItemEvent.Primary`.
**Forge and Fabric**: The click is detected using the swing arm packet. [Sponge](https://github.com/SpongePowered/Sponge/blob/api-8/src/mixins/java/org/spongepowered/common/mixin/core/server/network/ServerGamePacketListenerImplMixin.java#L435) and [Spigot](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/nms-patches/net/minecraft/server/network/PlayerConnection.patch#1157) are using this same technique. Sponge and Spigot are doing raytracing to ensure it's a click on air and not on a block but in our case we don't need to do raytracing, just deduplication.

There was a workaround on Forge where the mod if installed on the client would notify the server about the click. This workaround is no longer needed and has been removed.

### Right click on block

The right click on a block was handled twice because a RightClickBlock event is always followed by a RightClickAir event on all platforms. To fix this, Event handlers store the tick number during which the last interaction occured and ignore air interactions that are too close (tick delta <= 1).

This mechanism is active for both left and righ clicks.

### CODE REVIEW

**Comment by me4502** [comment]
Has the performance impact of this been measured? I remember previously doing this for CraftBook in the past, purely with sign clicks that would sometimes be duplicated and found that there was actually a measurable impact on some servers.

**Comment by Yeregorix** [comment]
I haven't measured the performance yet.

The deduplication is just a map, a specialized fastutil map, so it should have a really low cpu and memory footprint. In fact, the deduplication, would probably avoid some extra processing of double clicks so it will slightly improve performance on this side.

About the swing arm packet to detect a left click on air, this is what has been used by Bukkit and Sponge for years and afaik no one has ever reported a performance issue on this part. Moreover, the implementation here is much simpler and lighter (no raytracing needed).

Just in case, I will do some measures in-game with Spark's profiler but I doubt I will see anything noticeable.

**Review by me4502 — APPROVED** [review]

**Comment by Yeregorix** [comment]
I rebased on latest version/7.2.x commit.

I did some benchmarks in-game using Spark and an auto-clicker. 
World: superflat (to avoid measuring worldgen lag).
Click speed: 100 clicks / second (theoretical click speed is limited to 20c/s but it doesn't hurt to ensure extra packets will not cause overload).
Number of clicks: 3000 left clicks then 3000 right clicks.

Results:
Sponge: https://spark.lucko.me/ku7MJdFyW8
Fabric: https://spark.lucko.me/lxoD5XGsYX
Forge: https://spark.lucko.me/iiRgGkyhr1
Spigot: https://spark.lucko.me/OPUxcJbsqN

As expected the performance impact is negligible on all platforms. The benchmark procedure is manual and not rigorous but I doubt we would see a different result with another approach.

File: worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java [hunk 1]
```diff
@@ -86,6 +93,19 @@ public void onPlayerCommandSend(PlayerCommandSendEvent event) {
         );
     }
 
+    private static int getCurrentTick() {
+        if (PaperLib.isPaper()) {
+            return Bukkit.getCurrentTick();
+        }
+        return (int) (System.currentTimeMillis() / 50);
```
[the remark this answers is not in the document]
Comment (Yeregorix) [reply worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java:100]: It is not truly arbitrary. Since Spigot does not provide a way to get the current tick, I use the current time as an approximation. There is an offset between the real current tick and the approximation but we only want to detect when the delay between two clicks is too short so the offset is not a problem. Sure we can use `nanoTime`, this would improve accuracy a bit. I doubt it will have a significant impact but it cannot hurt.

[the remark this answers is not in the document]
Comment (octylFractal) [reply worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java:100]: I mainly want to avoid time going backwards, which could cause oddities.

File: worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java [hunk 2]
```diff
@@ -124,31 +144,36 @@ public void onPlayerInteract(PlayerInteractEvent event) {
             }
 
         } else if (action == Action.LEFT_CLICK_AIR) {
-
-            if (we.handleArmSwing(player)) {
+            if (!isDuplicateInteraction(player) && we.handleArmSwing(player)) {
```
[the remark this answers is not in the document]
Comment (Yeregorix) [reply worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java:147]: I used to deduplicate all kind of clicks but during tests I noticed that if I both left click and right click at the same time, I could break blocks while holding a wooden axe. The fix here is to never deduplicate `left click on block` events to ensure they are always cancelled when needed. 

We don't need to deduplicate `left click on block` events because they always occurs before the `click on air` events. As long as we deduplicate `click on air` events, this is fine.

Some events need to be deduplicated, some not, that's why the filter is not on the top.

File: worldedit-fabric/src/main/java/com/sk89q/worldedit/fabric/mixin/MixinServerGamePacketListenerImpl.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    private void onAction(ServerboundPlayerActionPacket packet, CallbackInfo ci) {
+        switch (packet.getAction()) {
+            case DROP_ITEM:
+            case DROP_ALL_ITEMS:
+            case START_DESTROY_BLOCK:
+                this.ignoreSwingPackets++;
+                break;
+            default:
+                break;
+        }
```
[the remark this answers is not in the document]
Comment (Yeregorix) [reply worldedit-fabric/src/main/java/com/sk89q/worldedit/fabric/mixin/MixinServerGamePacketListenerImpl.java:61]: I've been using Java 8 for too long. I will change that!

File: worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Yeregorix) [reply worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java:100]: Oh yes, I forgot that `currentTimeMillis` is not monotonic, while `nanoTime` is.

File: worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (octylFractal) [reply worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java:147]: Hmm, that makes me wonder if this is fully correct. Perhaps we should also cache the cancellation for the last result and repeat it, then?

**Comment by Yeregorix** [comment]
> Overall the way the debounce is handled is fine, but I don't like having the code repeated in every platform. Why not do this in `we.handleXClick` instead?

There is some platform specific stuff related to deduplication. 

LCA -> always deduplicated.
RCA -> always deduplicated.
LCB -> never deduplicated.
RCB -> deduplicated only for Bukkit. This is because the events fire in a slightly different order in Bukkit than other platforms.

It's been a while since I've worked on this, I will probably check again if the Bukkit-specific stuff is really needed.

> I also think that the fix for LCA in Forge/Fabric should be separated from this PR.

LCA is detected using the swing arm packet which is also present for some other click scenarios. In this case, LCA always happen last, so we must deduplicate it. Since LCA is heavily relying on deduplication I don't think we should move it into another PR. LCA without deduplication would trigger a lot of false positives.

File: worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yeregorix) [reply worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/WorldEditListener.java:147]: Yes that would make sense. I'll try and see what the implications of this are.

**Comment by Yeregorix** [comment]
I applied your suggestions!
- Rebased.
- Factorized the debouncing mechanism. *
- Modern switch expression.
- `nanoTime`

* I added a `getTickCount` method into `Platform`. I'm not sure whether this is the right place for it or not.
I have also discovered that field `server` was always `null` inside `ForgePlatform` so I fixed it.

**Comment by me4502** [comment]
Would it be possible to please rebase this due to conflicts? (involves 1.20 update too)

**Comment by Yeregorix** [comment]
Sure I will do ASAP but I'm really busy this week

**Comment by Yeregorix** [comment]
Rebased and retested.

File: worldedit-forge/src/main/java/com/sk89q/worldedit/forge/ForgePlatform.java [hunk 4]
```diff
@@ -120,12 +116,18 @@ public int schedule(long delay, long period, Runnable task) {
     @Override
     @Nullable
     public ForgeWatchdog getWatchdog() {
+        if (watchdog == null) {
```
Comment (octylFractal) [inline worldedit-forge/src/main/java/com/sk89q/worldedit/forge/ForgePlatform.java:119]: Why was this moved here? Doing this check over and over seems costly.

**Comment by me4502** [comment]
Thanks for this :)
